packages/react/src/hooks/program.ts

```diff
@@ -0,0 +1,200 @@
+  commitment?: CommitmentLevel | null;
+  instruction: TInstructionName;
+};
+
+type UseProgramQueryInput<
+  TAccounts extends Record<string, AccountFetcher<any>>,
+  TAccountName extends keyof TAccounts,
+> = Omit<UseQueryOptions<Awaited<ReturnType<TAccounts[TAccountName]>>, Error>, "queryFn" | "queryKey"> & {
+  account: TAccountName;
+  address: Address;
+  rpc: Parameters<TAccounts[TAccountName]>[0];
+};
+
+const CONFIRM_TRANSATION_MAX_RETRIES = 15;
+const CONFIRM_TRANSATION_COMMITMENT_FINALIZED_MILLISECOND_DELAY = 2000;
+const CONFIRM_TRANSATION_COMMITMENT_CONFIRMED_MILLISECOND_DELAY = 100;
+const CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY = 2000;
+
+
+async function waitForConfirmation(rpc: SolanaClient['rpc'], signature: Signature, commitment: CommitmentLevel): Promise<void> {
+  const initialDelay =
+    commitment === "finalized"
+      ? CONFIRM_TRANSATION_COMMITMENT_FINALIZED_MILLISECOND_DELAY
+      : CONFIRM_TRANSATION_COMMITMENT_CONFIRMED_MILLISECOND_DELAY;
+  let retryCount = 0;
+  let delay = initialDelay;
+
+  while (retryCount < CONFIRM_TRANSATION_MAX_RETRIES) {
+    try {
+      const response = await rpc.getSignatureStatuses([signature]).send();
+      const status = response.value?.[0];
+
+      if (status?.err) {
+        throw new Error(`Transaction failed: ${JSON.stringify(status.err)}`);
+      }
+
+      if (status?.confirmationStatus) {
+        const currentLevel = status.confirmationStatus;
+
+        if (currentLevel === commitment || (commitment === "confirmed" && currentLevel === "finalized")) {
+          return;
+        }
+      }
+
+      await new Promise((resolve) => setTimeout(resolve, delay));
+      delay = Math.min(delay * 1.2, CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY);
+      retryCount++;
+    } catch (error) {
+      await new Promise((resolve) => setTimeout(resolve, delay));
+      delay = Math.min(delay * 1.2, CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY);
+      retryCount++;
+      if (error instanceof Error && error.message.includes("Transaction failed")) {
+        throw error;
+      }
+    }
+  }
+}
+
+export function createProgramHook<
+  TInstructions extends Record<string, InstructionBuilder<any>>,
+  TAccounts extends Record<string, AccountFetcher<any>>,
+>(config: ProgramHookConfig<TInstructions, TAccounts>) {
+  const defaultCommitment = config.commitment === undefined ? "confirmed" : config.commitment;
+
+  function useProgramMutation<TInstructionName extends keyof TInstructions>(
+    input: UseProgramMutationInput<TInstructions, TInstructionName>,
+  ) {
+    const queryClient = useQueryClient();
+    const { instruction, commitment: mutationCommitment, ...options } = input;
+    const instructionFn = config.instructions[instruction];
+
+    return useMutation({
+      ...options,
+      mutationFn: async (input: {
+        commitment?: CommitmentLevel;
+        params: Parameters<TInstructions[TInstructionName]>[0];
+        rpc: SolanaClient['rpc'];
+        signAndSend: SignAndSendFn;
```

Comment:
These should be moved into program.ts for the next version but I think having them as input parameters works for now.